tests/test_smart_ptr.cpp

```diff
@@ -353,6 +363,15 @@ TEST_SUBMODULE(smart_ptr, m) {
     m.def("print_myobject3_3", [](const std::shared_ptr<MyObject3> &obj) { py::print(obj->toString()); });
     m.def("print_myobject3_4", [](const std::shared_ptr<MyObject3> *obj) { py::print((*obj)->toString()); });
 
+    py::class_<MyObject3a>(m, "MyObject3a");
+    m.def("make_myobject3_1", []() { return new MyObject3a(8); });
+    m.def("make_myobject3_2", []() { return std::make_shared<MyObject3a>(9); });
+    m.def("print_myobject3a_1", [](const MyObject3a *obj) { py::print(obj->toString()); });
+    m.def("print_myobject3a_2", [](std::shared_ptr<MyObject3a> obj) { py::print(obj->toString()); });
+    m.def("print_myobject3a_3", [](const std::shared_ptr<MyObject3a> &obj) { py::print(obj->toString()); });
+    // this doesn't compile, should it?
+    //m.def("print_myobject3a_4", [](const std::shared_ptr<MyObject3a> *obj) { py::print((*obj)->toString()); });
```

Comment:
No. A `shared_ptr` is expected to be passed by some kind of reference (or value).

Reply:
> No. A `shared_ptr` is expected to be passed by some kind of reference (or value).

I'm thinking of allowing const pointers, but definitely not non-const pointers.
Rationale: It could be useful to pass `nullptr`, that seems like a valid use case.

Reply:
A `nullptr` can be represented by an empty `shared_ptr`. IMO we always want to pass a pointer to a held object.
The `shared_ptr` is just a convenient replacement for a raw pointer. Passing a pointer to a (shared) pointer doesn't make sense to me - no matter being it const or not: We cannot modify the pointer anyway.

Reply:
> A `nullptr` can be represented by an empty `shared_ptr`. IMO we always want to pass a pointer to a held object.
> The `shared_ptr` is just a convenient replacement for a raw pointer. Passing a pointer to a (shared) pointer doesn't make sense to me - no matter being it const or not: We cannot modify the pointer anyway.

OK, let's keep this as-is for now. Let's just wait and see if anyone later has a strong need for it.